### ISSUE TITLE
Advice provider refactoring

### ISSUE THREAD

**Issue by bobbinth** [maticnetwork/miden#401]
After #393 is done, we will have refactored all IO operations except for the ones which deal with the advice tape. These operations are: `push.adv.n` and `loadw.adv`.

First, we should probably rename these operations to be consistent with the new naming conventions. Specifically:
* `loadw_adv` should be `adv_loadw`.
* Renaming `push_adv.n` is a bit more tricky since we use only load/store verbs in other place, but for the lack of a better option, I think we can replace it with `adv_push.n` (unless someone has better suggestions).

Second, a more fundamental issue with the advice provider is that we have only a single advice tape. For simple programs this tape can be pre-loaded with some values, and then the program can read these values one-by-one. However, for more complex program, pre-loading the tape would be far from trivial.

Imagine there is a program which needs to "un-hash" one of two values based on some condition which it needs to compute dynamically. To know which one of pre-images need to be put on the advice tape, we'd first need to execute the program up to the point the condition is computed, then once we know the condition, we could initialize the tape appropriately, and only after that, we could execute the program to the end. Needless to say that this approach is not workable for even moderately complicated programs.

One relatively simple way to address the above issue is to have many advice tapes. That is, we could replace a single tape in the advice provider with a key-value map where a key is some tape identifier and the value is the tape itself. This map would always have at least one tape (i.e., for key `0`), but could be initialized with any number of tapes under various keys.

To identify which tape to read values from, we could introduce a concept of an _active tape_. That is, only one tape can be active at a given time, and `adv_loadw` and `adv_push` instructions would always read from the active advice tape.

To change the active tape we could use a decorator - maybe something like `adv_config` (other name suggestions are welcome). The effect of this decorator would be set the active tape to the tape with the key equal to the top 4 items on the stack. The reason for using 4 elements as the key is because we frequently want to "un-hash" some value, and it is convenient to be able to look up hash pre-image by its hash.

A quick example. Let's say we have a Merkle tree. We want to get a leaf at position 1 from the tree, and then get hash pre-image of this leaf. Denoting leaf at position 1 as $n$, we assume that pre-image of $n$ is a tuple $a$, $b$. That is $n = hash(a, b)$. The program to do this could look like so:
```
begin
    push.1.2.3.4     # push Merkle root of the tree onto the stack
    push.1           # push node index onto the stack
    push.16          # push the depth of the tree onto the stack
    mtree.get        # load the node value onto the stack
    adv_config       # set the active tape to the tape with key equal to the node value
    swapw
    adv_loadw        # load the first 4 elements (a) from the advice tape onto the stack
    adv_push.4       # load another 4 elements (b) from the advice tape onto the stack
    repeat.8
        dup.7        # make a copy of the top 8 elements
    end
    rphash           # compute hash(a, b)
    swapw
    swapw.3
    eqw              # make sure hash(a, b) = n
    assert
    swapw            # at this point, the stack will look like [b, a, ...]
end
```
To run the above program, we need to initialize advice provider with a map of tapes where one of the entries is `(n, [a, b])` (i.e., the tape containing values `[a, b]` is under the key `n`.

Side note: the above program is rather cumbersome and we probably would want something better than that. My hope that methodology discussed in #336 will make it much more efficient, but having "tape maps" as described above would be very useful there too.

**Comment by frisitano** [comment]
Is the idea to initialise the advice provider map with advice tapes for all possible permutations that may be required by the conditional program or would the advice tapes be constructed dynamically as required?  I think the latter is a more powerful construct especially in the context of performing db lookups to fetch merkle inclusion proofs from a db backend which would serve as advise sets.

**Comment by bobbinth** [comment]
I think the initial implementation can be pretty simple and assume that all tapes are in memory. We can improve on this later with a more sophisticated implementation which can perform layered lookups (i.e., check memory first, and if not go to the database).

My first goal is to make sure the "front end" of the VM (i.e., the assembly language) is powerful enough to handle these use cases efficiently. Once this works, we can improve the advice provider without altering the front end (or at least that's the idea).

**Comment by bobbinth** [comment]
I've realized that adding multiple advice tapes to the advice provider may not cover all the use cases we want and that there might be a better way. The main problem with multiple advice tapes is that we are trying to simulate a key-value map, but it is a key-value map where value under each key can be read only once. If we wanted to read a value under the same key twice, the multi-tape approach would not work (at leas not as described).

So, if we want to have a key-value map of the form `Word -> Vec<Felt>`, why not just have this map and to provide an advice injector which would inject values under a given key to the front of a single tape? I think that would be simpler and more flexible approach.

So, instead of `adv_config` decorator, we'd have a new advice injector. Maybe something like `adv.keyval` (other name suggestions are welcome). This injector would work somewhat similarly to `adv_config`:
1. It would interpret the top 4 elements of the stack as a key,
2. Use this key to look up a value - i.e., a vector of filed elements - in the map,
3. But instead of setting an active advice tape, it would simply inject these elements at the front of the advice tape.

The map itself would need to be in the `AdviceProvider` and we'd also need to add a method to the advice provider to read values under a given key - maybe something like:

```Rust
pub fn get_values(&self, key: Word) -> Option<&[Felt]>
```

Then, when processing `adv.keyval` injector, we'd use this method to get the values for the key from the advice provider, and then would use `write_tape()` method to write them at the front of the advice tape (although, it might be good to add another method which could write multiple values at a time).

**Comment by frisitano** [comment]
Out of interest why does the multi-tape approach not work for multiple reads.  Is it because the tapes would be drained when they are read the first time?

I wonder if we should introduce an additional field into the advice provider in the from on an `AdviceProviderBackend` such we have:

```rust
pub struct AdviceProvider {
    step: u32,
    tape: Vec<Felt>,
    sets: BTreeMap<[u8; 32], AdviceSet>,
    backend: AdviceProviderBackend,
}

pub trait AdviceProviderBackend {
  fn get_values(&self, key: Word) -> Option<&[Felt]>;
  
  fn get_inclusion_proof(&self, key: Word) -> Option<AdviceSet>;
  
  ...
```

It would then be the responsibility of the `AdviceProvider` to call into the backend to retrieve data and populate the tape with the data.  In the first instance an in-memory backend is probably sufficient however this pattern could be extended to work over database backends in the future.  It also provides a good level of abstraction such that the backend can be tailored to the platform miden is deployed on.

**Comment by bobbinth** [comment]
> Out of interest why does the multi-tape approach not work for multiple reads. Is it because the tapes would be drained when they are read the first time?

Yes, exactly. When we read an element, it is removed from the tape.

> I wonder if we should introduce an additional field into the advice provider in the from on an `AdviceProviderBackend`

I was thinking about something like this myself, though, in my mind it would probably make sense to make the whole advice provider a trait. This trait could look like:
```Rust
pub trait AdviceProvider {
    
    fn get_values(&self, key: Word) -> Option<&[Felt]>;

    fn read_tape(&mut self) -> Option<Felt>;

    fn write_tape(&mut self, value: Felt);
    fn write_tape_many(&mut self, values: &[Felt]);

    fn get_tree_node(&self, root: Word, depth: Felt, index: Felt) -> Result<Word, ExecutionError>;
    fn get_merkle_path(&self, root: Word, depth: Felt, index: Felt) -> Result<Vec<Word>, ExecutionError>;
    fn update_merkle_leaf(&mut self, root: Word, index: Felt, leaf_value: Word, update_in_copy: bool) -> Result<Vec<Word>, ExecutionError>;

    fn advance_clock(&mut self);
}
```

Then, when we need to generate a proof, we could pass any implementation of an advice provider to the `prove()` function, which could have the following interface (more on this is in #409):
```Rust
pub fn prove<A: AdviceProvider>(
    program: &Program,
    stack_inputs: &[u64],
    advice_provider: &mut A,
    options: &ProofOptions,
) -> Result<(ProgramOutputs, StarkProof), ExecutionError>
```
We could provide a default, memory-based implementation of the advice provider, but people would be able to replace it with w/e better works for them (including DB-backed ones).

**Comment by frisitano** [comment]
Sounds good to me :+1:

**Comment by bobbinth** [comment]
Closed #406. The discussion about converting advice provider into a trait is moved to #409
